File: intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/plugin/PluginXmlXIncludePathResolver.java
package com.jetbrains.plugin.structure.intellij.plugin;

import com.jetbrains.plugin.structure.intellij.utils.URLUtil;
import com.jetbrains.plugin.structure.intellij.utils.xincludes.DefaultXIncludePathResolver;
import com.jetbrains.plugin.structure.intellij.utils.xincludes.XIncludeException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Sergey Patrikeev
 */
public class PluginXmlXIncludePathResolver extends DefaultXIncludePathResolver {

  private final List<URL> myPluginMetaInfUrls;

  public PluginXmlXIncludePathResolver(List<URL> metaInfUrl) {
    myPluginMetaInfUrls = new ArrayList<URL>(metaInfUrl);
  }

  @NotNull
  private URL defaultResolve(@NotNull String relativePath, @Nullable String base) {
    if (base != null && relativePath.startsWith("/META-INF/")) {
      //for plugin descriptor the root is a directory containing the META-INF
      try {
        return new URL(new URL(base), ".." + relativePath);
      } catch (MalformedURLException e) {
        throw new XIncludeException(e);
      }
    }
    return super.resolvePath(relativePath, base);
  }

  @NotNull
  private URL getMetaInfRelativeUrl(@NotNull URL metaInf, @NotNull String relativePath) throws MalformedURLException {
    if (relativePath.startsWith("/")) {
      return new URL(metaInf, ".." + relativePath);
    } else {
      return new URL(metaInf, relativePath);
    }
  }

  @NotNull
  @Override
  public URL resolvePath(@NotNull String relativePath, @Nullable String base) {
    URL url = defaultResolve(relativePath, base);
    if (!URLUtil.resourceExists(url)) {
      for (URL metaInf : myPluginMetaInfUrls) {
        try {
          URL entryUrl = getMetaInfRelativeUrl(metaInf, relativePath);
          if (URLUtil.resourceExists(entryUrl)) {
            return entryUrl;
          }
        } catch (MalformedURLException ignored) {
        }
      }
    }
    return url;
  }
}

File: intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/plugin/PluginXmlXIncludePathResolver.kt
package com.jetbrains.plugin.structure.intellij.plugin

import com.jetbrains.plugin.structure.base.utils.isJar
import com.jetbrains.plugin.structure.base.utils.isZip
import com.jetbrains.plugin.structure.intellij.utils.ThreeState
import com.jetbrains.plugin.structure.intellij.utils.URLUtil
import com.jetbrains.plugin.structure.intellij.utils.xincludes.DefaultXIncludePathResolver
import com.jetbrains.plugin.structure.intellij.utils.xincludes.XIncludePathResolver
import java.io.File
import java.net.MalformedURLException
import java.net.URL

class PluginXmlXIncludePathResolver(files: List<File>) : XIncludePathResolver {

  private val metaInfUrls = getMetaInfUrls(files)

  private fun getMetaInfUrls(files: List<File>) =
      files.asSequence()
          .filter { it.isJar() || it.isZip() }
          .mapNotNull {
            try {
              URLUtil.getJarEntryURL(it, "${IdePluginManager.META_INF}/")
            } catch (e: MalformedURLException) {
              null
            }
          }.toList()

  private fun getRelativeUrl(base: URL, path: String) =
      if (path.startsWith("/")) {
        URL(base, "..$path")
      } else {
        URL(base, path)
      }

  override fun resolvePath(relativePath: String, base: String?): URL =
      metaInfUrls.asSequence()
          .mapNotNull {
            try {
              getRelativeUrl(it, relativePath)
            } catch (e: MalformedURLException) {
              null
            }
          }
          .find { URLUtil.resourceExists(it) == ThreeState.YES }
          ?: DefaultXIncludePathResolver.INSTANCE.resolvePath(relativePath, base)
}
